packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+    let edges = [];
+    let typeMap = this.#typeMap || (this.#typeMap = new TypeMap(this));
+    for (let [type, nodes] of typeMap.getConnectedFrom(from)) {
+      for (let to of nodes) {
+        edges.push({type: (type: any), to});
+      }
+    }
+    return edges;
+  }
+
+  /**
+   *
+   */
+  getEdges(
+    from: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): $ReadOnlySet<NodeId> {
+    return new Set(this.getNodesConnectedFrom(from, type));
+  }
+
+  /**
+   * Get the list of nodes connected from this node.
+   */
+  getNodesConnectedFrom(
+    from: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): NodeId[] {
+    let typeMap = this.#typeMap || (this.#typeMap = new TypeMap(this));
+    let isAllEdgeTypes =
+      type === ALL_EDGE_TYPES ||
+      (Array.isArray(type) && type.includes(ALL_EDGE_TYPES));
+
+    let nodes = [];
+    if (isAllEdgeTypes) {
+      for (let toSet of typeMap.getConnectedFrom(from).values()) {
+        nodes.push(...toSet);
+      }
+    } else if (Array.isArray(type)) {
+      for (let typeNum of type) {
+        nodes.push(...typeMap.getConnectedFromType(from, typeNum));
+      }
+    } else {
+      nodes.push(...typeMap.getConnectedFromType(from, (type: any)));
+    }
+    return nodes;
+  }
+
+  /**
+   * Get the list of nodes connected to this node.
+   */
+  getNodesConnectedTo(
+    to: NodeId,
+    type:
+      | AllEdgeTypes
+      | TEdgeType
+      | NullEdgeType
+      | Array<TEdgeType | NullEdgeType> = 1,
+  ): NodeId[] {
+    let typeMap = this.#typeMap || (this.#typeMap = new TypeMap(this));
+    let isAllEdgeTypes =
+      type === ALL_EDGE_TYPES ||
+      (Array.isArray(type) && type.includes(ALL_EDGE_TYPES));
+
+    let nodes = [];
+    if (isAllEdgeTypes) {
+      for (let fromSet of typeMap.getConnectedTo(to).values()) {
+        nodes.push(...fromSet);
+      }
+    } else if (Array.isArray(type)) {
+      for (let typeNum of type) {
+        nodes.push(...typeMap.getConnectedToType(to, typeNum));
+      }
+    } else {
+      nodes.push(...typeMap.getConnectedToType(to, (type: any)));
+    }
+    return nodes;
+  }
+
+  /**
+   * Create a hash of the edge connecting the `from` and `to` nodes.
+   *
+   * This hash is used to index the edge in the `edges` array.
+   *
+   */
+  hash(from: NodeId, to: NodeId, type: TEdgeType | NullEdgeType): EdgeHash {
+    // A crude multiplicative hash, in 3 steps:
+    // 1. Serialize the args into an integer that reflects the argument order,
+    // shifting the magnitude of each argument by the sum
+    // of the significant digits of the following arguments,
+    // .e.g., `hash(10, 24, 4) => 10244`.
+    // $FlowFixMe[unsafe-addition]
+    // $FlowFixMe[incompatible-type]
+    let hash = '' + from + to + type - 0;
```

Comment:
This seems like it could be a perf issue. Converting numbers to strings and back is expensive and this is a very hot function.

Also, your hash32shift function only works on 32 bit integers, and this could easily create a number larger than that. Your node ids are up to 32 bits a piece already and you are concatenating them here.

Did you try some kind of hash combining function? No idea how good this on is but it might be faster?

```js
let hash = 17;
hash = hash * 37 + hash32shift(from);
hash = hash * 37 + hash32shift(to);
hash = hash * 37 + hash32shift(type);
```

Reply:
Very interesting! We thought string conversion seemed like a not great idea, but it was never indicated as a bottleneck in our profiling, so we left it. However, I've done some microbenching with your suggested impl, and it is indeed faster, but does not appear to degrade uniformity at all, which seems like a no-brainer:

'hash' is string conversion
'hash2' is prime multiplication

163,633 nodes, 1,351,038 edges:

|       | ops/sec | max collisions | avg collisions |
| :---- | ------: | -------------: | -------------: |
| hash  | 505,108 |              8 |           0.46 |
| hash2 | 884,441 |              9 |           0.46 |

92,421 nodes, 255,538 edges:

|       |   ops/sec | max collisions | avg collisions |
| :---- | --------: | -------------: | -------------: |
| hash  |   924,997 |              5 |           0.34 |
| hash2 | 1,384,921 |              7 |           0.34 |